jormungandr/src/blockchain/tip.rs

```diff
@@ -202,52 +202,69 @@ impl TipUpdater {
         Ok(())
     }
 
-    /// this function will re-process the tip against the different branches
+    /// this function will re-process the tip against the different branches.
     /// this is because a branch may have become more interesting with time
     /// moving forward and branches may have been dismissed
-    async fn reprocess_tip(&mut self) -> Result<(), Error> {
-        let branches: Vec<Arc<Ref>> = self.blockchain.branches().branches().await;
-        let tip_as_ref = self.tip.get_ref().await;
-
-        let others = branches
-            .iter()
-            .filter(|r| !Arc::ptr_eq(r, &tip_as_ref))
-            .collect::<Vec<_>>();
+    #[instrument(level = "debug", skip_all, fields(current_tip = %tip.header().description()))]
+    async fn reprocess_tip(
+        blockchain: Blockchain,
+        tip: Arc<Ref>,
+        mut mbox: MessageBox<Arc<Ref>>,
+    ) -> Result<(), Error> {
+        use std::cmp::Ordering;
+        let branches = blockchain.branches().await?;
+        let storage = blockchain.storage();
+
+        let best_branch = branches.into_iter().map(Branch::into_ref).max_by(|a, b| {
+            match chain_selection::compare_against(storage, &a, &b) {
+                ComparisonResult::PreferCurrent => Ordering::Greater,
+                ComparisonResult::PreferCandidate => Ordering::Less,
+            }
+        });
 
-        for other in others {
-            self.process_new_ref(Arc::clone(other)).await?
+        if let Some(new_tip) = best_branch {
+            if !Arc::ptr_eq(&tip, &new_tip) {
```

Comment:
I wonder if this is truly necessary. But I can't imagine a case where it should case problems, so I guess it's fine (except maybe some `RefCache` refactors?)

Reply:
It's mostly to provide better clarity in the logs on what's happening

Reply:
I meant the `ptr_eq` instead of using normal equality, but I thought it was a the tip id and not a `Ref`, and I guess that's not comparable, so please ignore this.